sdk/src/main/java/com/google/cloud/dataflow/sdk/util/common/Counter.java

```diff
@@ -418,17 +452,191 @@ public boolean isCompatibleWith(Counter<?> that) {
 
   //////////////////////////////////////////////////////////////////////////////
 
-  /** The name of this counter. */
-  protected final String name;
-
   /** The kind of aggregation function to apply to this counter. */
   protected final AggregationKind kind;
 
   /** The commit state of this counter. **/
   protected final AtomicReference<CommitState> commitState;
 
+  /** which name to use from the original context. */
+  public static enum NameType {
+    /** Counter does not have a structured name. */
+    UNSTRUCTURED,
+
+    /** Structured name should use the original name. */
+    ORIGINAL,
+
+    /** Structured name should use the optimized name. */
+    OPTIMIZED
+  }
+
+  /** A hashable name structure which will uniquely identify a counter. */
+  public interface Name {
+    public String getName();
+    public NameContext getContext();
+    public NameType getType();
+  }
+
+  /** A StructuredName which uses the Original step name from the NameContext. */
+  public static class StructuredNameOriginal implements Name {
+
+    /** The name of this counter. */
+    private final String name;
+
+    /** The context this counter was named in.  */
+    private final NameContext context;
+
+    public StructuredNameOriginal(String name, NameContext context) {
+      this.name = name;
+      this.context = context;
+    }
+
+    @Override
+    public String getName() {
+      return this.name;
+    }
+
+    @Override
+    public NameContext getContext() {
+      return this.context;
+    }
+
+    @Override
+    public NameType getType() {
+      return NameType.ORIGINAL;
+    }
+
+    @Override
+    public String toString() {
+      StringBuilder sb = new StringBuilder();
+      sb.append("name: " + name + ", ");
+      sb.append("context: " + context.toString() + ", ");
+      sb.append("nameType: Original");
+      return sb.toString();
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(name, context.getOriginalName());
+    }
+
+    @Override
+    public boolean equals(Object other) {
+      if (other instanceof StructuredNameOriginal) {
+        StructuredNameOriginal that = (StructuredNameOriginal) other;
+        return this.name.equals(that.name)
+            && this.context.getOriginalName().equals(that.context.getOriginalName());
+      }
+      return false;
+    }
+  }
+
+  /** A StructuredName which uses the Optimized step name from the NameContext. */
+  public static class StructuredNameOptimized implements Name {
```

Comment:
This could use `@AutoValue` to implement most of the logic (eg., all the getters, the toString, the hashCode, the equals, etc.)